packages/valory/skills/decision_maker_abci/behaviours/base.py

```diff
@@ -725,6 +766,150 @@ def _write_benchmark_results(
             row = ",".join(results_text) + NEW_LINE
             results_file.write(row)
 
+    def _calc_token_amount(
+        self,
+        operation: TradingOperation,
+        amount_field: str,
+        amount_param_name: str,
+        amount_param_value: int,
+    ) -> WaitableConditionType:
+        """Calculate the token amount for buying/selling."""
+
+        response_msg = yield from self.get_contract_api_response(
+            performative=ContractApiMessage.Performative.GET_RAW_TRANSACTION,  # type: ignore
+            contract_address=self.market_maker_contract_address,
+            contract_id=str(FixedProductMarketMakerContract.contract_id),
+            contract_callable=f"calc_{operation.value}_amount",
+            outcome_index=self.outcome_index,
+            chain_id=self.params.mech_chain_id,
+            **{
+                amount_param_name: amount_param_value,
+            },  # type: ignore
+        )
+        if response_msg.performative != ContractApiMessage.Performative.RAW_TRANSACTION:
+            self.context.logger.error(
+                f"Could not calculate the {operation} amount: {response_msg}"
+            )
+            return False
+
+        token_amount = response_msg.raw_transaction.body.get(amount_field, None)
+        if token_amount is None:
+            self.context.logger.error(
+                f"Something went wrong while trying to get the {amount_field} for the conditional token: {response_msg}"
+            )
+            return False
+
+        if operation == TradingOperation.BUY:
+            self.context.logger.info(f"Function call buy amount: {token_amount}")
+            self.buy_amount = remove_fraction_wei(token_amount, self.params.slippage)
+        else:
+            self.context.logger.info(f"Function call sell amount: {token_amount}")
+            self.sell_amount = remove_fraction_wei(token_amount, self.params.slippage)
+        return True
+
+    def _calc_buy_amount(self) -> WaitableConditionType:
+        """Calculate the buy amount of the conditional token."""
+        return self._calc_token_amount(
+            operation=TradingOperation.BUY,
+            amount_field="amount",
+            amount_param_name="investment_amount",
+            amount_param_value=self.investment_amount,
+        )
+
+    def _calc_sell_amount(self) -> WaitableConditionType:
+        """Calculate the sell amount of the conditional token."""
+        return self._calc_token_amount(
+            operation=TradingOperation.SELL,
+            amount_field="amount",
+            amount_param_name="return_amount",
+            amount_param_value=self.investment_amount,
+        )
+
+    def _build_token_tx(self, operation: str) -> WaitableConditionType:
```

Comment:
Let's pass the operation as an enum instead to avoid human error and enhance readability.